src/socks5proxy/proxycontroller.cpp

```diff
@@ -0,0 +1,145 @@
+}
+
+void ProxyController::start() {
+  auto loopholeFile = QFileInfo(binaryPath());
+  if (!loopholeFile.exists()) {
+    Q_ASSERT(false);
+    return;
+  }
+
+  if (mCurrentProcess) {
+    mCurrentProcess->kill();
+    mCurrentProcess->deleteLater();
+  }
+  auto const url = generateProxyUrl();
+
+  mCurrentProcess = new QProcess();
+  mCrashSignal = QObject::connect(
+      mCurrentProcess, &QProcess::finished,
+      [this, url](int exitCode, QProcess::ExitStatus) {
+        logger.error() << "Loophole Closed Unexpected with " << exitCode;
+        logger.debug() << "Loophole scheudled restart";
+        mCurrentProcess->deleteLater();
+        start();
+      });
+  QObject::connect(mCurrentProcess, &QProcess::readyReadStandardOutput,
+                   [this]() {
+                     logger.debug() << "LOOPHOLE.exe"
+                                    << mCurrentProcess->readAllStandardOutput();
+                   });
+  QObject::connect(mCurrentProcess, &QProcess::started, [this, url]() {
+    logger.debug() << "Loophole available under:" << url.toString();
+    m_state.setValue(Started{url});
+  });
+  QObject::connect(mCurrentProcess, &QProcess::errorOccurred,
+                   [this](QProcess::ProcessError error) {
+                     switch (error) {
+                       case QProcess::ProcessError::Crashed:
+                         logger.error() << "Loophole crashed!";
+                         break;
+                       case QProcess::ProcessError::FailedToStart:
+                         logger.error() << "Loophole FailedToStart!";
+                         break;
+                       case QProcess::ProcessError::ReadError:
+                         logger.error() << "Loophole ReadError!";
+                         break;
+                       case QProcess::ProcessError::Timedout:
+                         logger.error() << "Loophole Timedout!";
+                         break;
+                       case QProcess::ProcessError::UnknownError:
+                         logger.error() << "Loophole UnknownError!";
+                         break;
+                       case QProcess::ProcessError::WriteError:
+                         logger.error() << "Loophole UnknownError!";
+                         break;
+                     }
+                     stop();
+                   });
+  mCurrentProcess->setProgram(loopholeFile.absoluteFilePath());
+  mCurrentProcess->setArguments(getArguments(url));
+  logger.error() << "Try to start Loophole!  "
+                 << loopholeFile.absoluteFilePath() << getArguments(url);
+  mCurrentProcess->start(QIODeviceBase::ReadOnly);
+}
+
+void ProxyController::stop() {
+  if (!mCurrentProcess && std::holds_alternative<Stopped>(m_state.value())) {
+    return;
+  }
+  logger.debug() << "Loophole stopping";
+  // We no longer need signals from this obj
+  m_state = Stopped{};
+  mCurrentProcess->kill();
+}
+
+bool ProxyController::canActivate() {

```

Comment:
I do not think this is a good check. You can have multiple versions of firefox, and each can have multiple profiles, and it's not required for the profile dir to be in a speficic folder, so this could go wrong, given we'd disable a feature if we fail to detect the extension.